backend/src/payments/services/payments.service.ts

```diff
@@ -0,0 +1,122 @@
+import {
+  ForbiddenException,
+  Injectable,
+  Logger,
+  RawBodyRequest,
+  UnauthorizedException,
+} from '@nestjs/common'
+import { ConfigService } from '@nestjs/config'
+import { Request } from 'express'
+import { Schema } from 'mongoose'
+import Stripe from 'stripe'
+import { UserService } from '../../db/services/user.service'
+import { PaymentOptions, paymentPlans } from '../interfaces/payments'
+import { StripeService } from './stripe.service'
+
+@Injectable()
+export class PaymentsService {
+  private logger = new Logger(PaymentsService.name)
+  private stripe: Stripe
+  constructor(
+    private stripeService: StripeService,
+    private configService: ConfigService,
+    private readonly userService: UserService,
+  ) {}
+
+  async createCheckoutSession(plan: string, userId: Schema.Types.ObjectId) {
+    const user = await this.userService.findOneById(userId)
+    if (!user)
+      throw new UnauthorizedException(
+        'This user does not exist and cannot make a purchase', // How unfortunate, we always want money
+      )
+
+    if (plan === user.paymentPlan)
+      throw new ForbiddenException('You cannot rebuy a plan') // Actually I would love to allow it if it means more money
+    if (paymentPlans[plan] < paymentPlans[user.paymentPlan]) {
+      throw new ForbiddenException('You cannot downgrade your plan') // Actually I would love to allow it if it means more money
+    }
+
+    let price_id: string
+
+    if (plan === 'single')
+      price_id = this.configService.get('STRIPE_ITEM_SINGLE')
+    if (plan === 'family')
+      price_id =
+        user.paymentPlan === 'single'
+          ? this.configService.get('STRIPE_ITEM_SINGLE_TO_FAMILY')
+          : this.configService.get('STRIPE_ITEM_FAMILY')
+
+    let customer = user.stripeCustomerId
+    if (!customer) {
+      customer = (await this.stripeService.customer_create(user.email)).id
+      await this.userService.updateUserStripeCustomerId(
+        user._id.toString(),
+        customer,
+      )
+    }
+
+    const session = await this.stripeService.checkout_session_create(
+      plan,
+      price_id,
+      customer,
+    )
+
+    this.logger.debug('customer:', customer)
+    await this.userService.updateUserCheckoutInformation(customer, {
```

Comment:
If this fails the session is never send to the user...does that make sense or can this lead to issues with stripe?

Reply:
I want to guarantee that our db contains data regarding a pending checkout, there is no issue in creating multiple on Stripe site since they just expire